refactor: executor functional options

pd93 opened the pull request:
This PR migrates the `Executor` to use the functional options pattern. This follows other similar changes to `taskfile.Reader` and `taskfile.Snippet`. This allows package API users to easily construct an `Executor` with sensible defaults and only supply a list of things they want to override.

Our internal CLI implementation has a special functional option in `internal/flags` called `WithFlags` which will gather all the CLI flags and pass them in as functional options.

vmaerten reviewed (approved):
Good refactor as always!

andreynering commented on executor.go:
```diff
@@ -0,0 +1,253 @@
+	}
+)
+
+func NewExecutor(opts ...ExecutorOption) *Executor {
```
Each of the functions in this file should have a documentation comment, do make them look pretty at https://pkg.go.dev/github.com/go-task/task/v3.

andreynering commented on executor.go:
```diff
@@ -0,0 +1,253 @@
+	}
+}
+
+func WithStdin(stdin io.Reader) ExecutorOption {
+	return func(e *Executor) {
+		e.Stdin = stdin
+	}
+}
+
+func WithStdout(stdout io.Writer) ExecutorOption {
+	return func(e *Executor) {
+		e.Stdout = stdout
+	}
+}
+
+func WithStderr(stderr io.Writer) ExecutorOption {
+	return func(e *Executor) {
+		e.Stderr = stderr
+	}
+}
```
In theory we could have an additional function that accepts all three, for shortness. Optional suggestion, though.

```go
func WithIO(stdin io.Reader, stdout, stderr io.Writer) ExecutorOption {
	return func(e *Executor) {
		e.Stdin = stdin
		e.Stdout = stdout
		e.Stderr = stderr
	}
}
```

pd93 replied:
To me, it makes more sense to keep these separate if you're passing different values into each of them. Sometimes you might only want to pass in 1 or 2 of them and leave the others blank. However, I think if you're setting them all to the same value, we could have something like this:

```go
func WithIO(rw io.ReadWriter) ExecutorOption {
	return func(e *Executor) {
		e.Stdin = rw
		e.Stdout = rw
		e.Stderr = rw
	}
}
```

This would be super handy in the new tests I've been writing where we want to write everything to a single buffer.

andreynering reviewed (approved):
Great work!

![Claps](https://github.com/user-attachments/assets/0da784bf-4a24-484a-bb66-4a0ffda4c01f)

pd93 commented:
Couple of minor changes:

- Added the missing `ExecutorWithVersionCheck` functional option in `cmd/task`
- Added prefixes to the functional options like in `taskfile.Snippet`
  - This helps with any potential naming conflicts in the future.
- A bunch of docstrings for everything.
